QueueDown/MultiStagePipes.cs

```diff
@@ -0,0 +1,146 @@
+                var producerPipe = new Pipe(new(pool));
+                readers.AddPipe(producerPipe.Reader);
+
+
+                // Write to the pipe
+                while (true) {
+                    var buffer = producerPipe.Writer.GetMemory();
+                    buffer.Span.Fill((byte)i);
+                    producerPipe.Writer.Advance(buffer.Length);
+                    await producerPipe.Writer.FlushAsync();
+                }
+            });
+        }
+
+        var _ = Task.Run(() => ReadAllPipes(readers, outputPipe.Writer));
+    }
+
+    /// <summary>
+    /// Reads from pipes in a round-robin manner and pushes output into final pipe
+    /// </summary>
+    private static async Task ReadAllPipes(PipeGroup stage1Pipes, PipeWriter outputPipe)
+    {
+        while (true) {
+            var pipes = stage1Pipes.Readers; //fresh read incase this changed
+            bool anyDataThisLoop = false;
+
+            foreach (var inputPipe in pipes) 
+            {
+                if (inputPipe.TryRead(out var result)) 
+                {
+                    anyDataThisLoop = true;
+
+                    await ProcessPipeResult(stage1Pipes, inputPipe, result, outputPipe);
+                }
+            }
+
+            if(!anyDataThisLoop)
+            {
+                //take the slow path if we are starved
+                var results = await stage1Pipes.WaitForMoreData();
+
+                foreach(var result in results)
+                {
+                    await ProcessPipeResult(stage1Pipes, result.Item1, result.Item2, outputPipe);
+                }
+            }
+        }
+    }
+
+    private static async Task ProcessPipeResult(PipeGroup group, PipeReader inputPipe, ReadResult result, PipeWriter outputPipe)
+    {
+        var inputBuffer = result.Buffer;
+
+        if (inputBuffer.IsSingleSegment) 
+        {
+            outputPipe.Write(result.Buffer.FirstSpan);
+        } 
+        else 
+        {
+            foreach (var segment in result.Buffer) 
+            {
+                outputPipe.Write(segment.Span);
+            }
+        }
+
+        var flushResult = await outputPipe.FlushAsync();
+
+        inputPipe.AdvanceTo(inputBuffer.End);
+
+
+        if (result.IsCompleted || flushResult.IsCompleted) {
+            group.RemovePipe(inputPipe);
+        }
+    }
+
+    
+
+    /// <summary>
+    /// Provides a way to add and remove pipes in a fast/thread safe way
+    /// </summary>
+    public class PipeGroup
+    {
+        private ImmutableArray<PipeReader> _readers = ImmutableArray<PipeReader>.Empty;
+
+        public ImmutableArray<PipeReader> Readers => _readers;
+
+
+        public void AddPipe(PipeReader reader)
+        {
+            ImmutableInterlocked.Update(ref _readers, static (array, arg) => array.Add(arg), reader);
+        }
+
+        public void RemovePipe(PipeReader reader)
+        {
+            ImmutableInterlocked.Update(ref _readers, static (array, arg) => array.Remove(arg), reader);
```

Comment:
This is O(N) right?

Reply:
Yes, I could have used an Immutable list or a perhaps a simple lock or something to guards a mutable structure, but I figured attaching and detaching the sources would be relatively infrequent and the N would be relatively low.

Reply:
That's fair. For short lived requests, you'd end up adding and remove it a lot.

Reply:
Hmm, good point, I kinda like the idea that this structure can be used to quickly attach and detach input pipes. I will tweak the design to make this cheaper.